segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/mocks/InMemoryChunkStorage.java

```diff
@@ -0,0 +1,295 @@
+                .build();
+    }
+
+    @Override
+    protected ChunkHandle doCreate(String chunkName) throws ChunkStorageException, IllegalArgumentException {
+        Preconditions.checkNotNull(chunkName);
+        if (null != chunks.putIfAbsent(chunkName, new InMemoryChunk(chunkName))) {
+            throw new ChunkAlreadyExistsException(chunkName, "InMemoryChunkStorage::doCreate");
+        }
+        return new ChunkHandle(chunkName, false);
+    }
+
+    @Override
+    protected boolean checkExist(String chunkName) throws ChunkStorageException, IllegalArgumentException {
+        return chunks.containsKey(chunkName);
+    }
+
+    @Override
+    protected void doDelete(ChunkHandle handle) throws ChunkStorageException, IllegalArgumentException {
+        InMemoryChunk chunk = getInMemoryChunk(handle);
+        if (null == chunk) {
+            throw new ChunkNotFoundException(handle.getChunkName(), "InMemoryChunkStorage::doDelete");
+        }
+        if (chunk.isReadOnly) {
+            throw new ChunkStorageException(handle.getChunkName(), "chunk is readonly");
+        }
+        chunks.remove(handle.getChunkName());
+    }
+
+    @Override
+    protected ChunkHandle doOpenRead(String chunkName) throws ChunkStorageException, IllegalArgumentException {
+        if (chunks.containsKey(chunkName)) {
+            return new ChunkHandle(chunkName, true);
+        }
+        throw new ChunkNotFoundException(chunkName, "InMemoryChunkStorage::doOpenRead");
+    }
+
+    @Override
+    protected ChunkHandle doOpenWrite(String chunkName) throws ChunkStorageException, IllegalArgumentException {
+        if (chunks.containsKey(chunkName)) {
+            return new ChunkHandle(chunkName, false);
+        }
+        throw new ChunkNotFoundException(chunkName, "InMemoryChunkStorage::doOpenWrite");
+    }
+
+    @Override
+    protected int doRead(ChunkHandle handle, long fromOffset, int length, byte[] buffer, int bufferOffset) throws ChunkStorageException, NullPointerException, IndexOutOfBoundsException {
+        InMemoryChunk chunk = getInMemoryChunk(handle);
+        if (fromOffset >= chunk.getLength()) {
+            throw new ChunkStorageException(handle.getChunkName(), "Attempt to read at wrong offset");
+        }
+        if (length == 0) {
+            throw new ChunkStorageException(handle.getChunkName(), "Attempt to read 0 bytes");
+        }
+
+        // This is implemented this way to simulate possibility of partial read.
+        InMemoryChunkData matchingData = null;
+
+        // TODO : This is OK for now. This is just test code, but need binary search here.
+        for (InMemoryChunkData data : chunk.inMemoryChunkDataList) {
+            if (data.start <= fromOffset && data.start + data.length > fromOffset) {
+                matchingData = data;
+                break;
+
+            }
+        }
+
+        if (null != matchingData) {
+            int startIndex = (int) (fromOffset - matchingData.start);
```

Comment:
we could move this block into a method in `InMemoryChunkData`

Reply:
This the only thing that `doRead` really does and the method is already small enough.

Reply:
I am telling this just to have a cleaner encapsulation for InMemoryChuckData